### PR TITLE
A hypothesis test which loads object template from json file and create hypothesis object

### PR DESCRIPTION
**Pull request by ke-zhang-rd**
Hypothesis object created by this process based on `container.json` will look like
```
Container(contents={Sample(composition='', description='', name='\U000e4a00\U00064a2a\t\U0008e662\n', projects=[], tags=['\U00056c15\U00086a09𗊚\U000ecaa6\x14\U000f3e75\x00\U001036f9\U000423dd\U000f195a\x06\r\x11\U0001a150\U0003f389\U0010ce9b\x15\x12\U00092be8#\U000f6d39', '\U000fb6fc\x1a\U0010c04e\x14\U000382c8']): 'LOCATION', Sample(composition='', description='', name='\x07#\U000847f8\U000f6578', projects=[], tags=[]): 'LOCATION', Sample(composition='\U00060b73/"', description='\U00046df7\x1d\U000749df(\x1f', name='/ \U000f49a4\U0008d7cf', projects=[], tags=[]): 'LOCATION', Sample(composition='', description='', name='\x17\x1c\x17\x19\x1b', projects=[], tags=['#\x1b\U000be9ef\U0009b5b6', '\x1e\U0007003a', '\U00086d1c\U00107cc6)-.\x1d\U00064fa4', '\U000a42f2\x14\U0007279e', '', '+', '', '\U00107542', '&%""\U000bcb76!\U0001ed63\U000385a0#\U00058815\x06!\U000fd3c0\x01\x1c,𐑄.\x07\U0010f0d4', '\x1a⮊', '', '\U000e7bfd\x16\U0003909e', '\U000cf4bc', '*\x08\x04\x16.\U000f1290\x11\x04ᾈ\x12\U0010d905', '\t\x1f']): 'LOCATION', Sample(composition='', description='', name='\x11\U0003eeaf\U0005c7b7\U0005a192', projects=[], tags=[]): 'LOCATION'}, kind='\U00109223𥉽', name='!! \x15-')
```

This also could be a beginning point for who want to try using hypothesis test/package later.

### CODE REVIEW

File: amostra/schemas/container.json [hunk 1]
```diff
@@ -21,10 +21,13 @@
             "description": "Maps a Sample Foreign key to a location"
         }
     },
+    "additionalProperties": false,
     "required": [
         "uuid",
         "revision",
-        "name"
+        "name",
+        "kind",
```
Comment (danielballan) [inline amostra/schemas/container.json]: Why does this testing PR have to change what is required?

File: amostra/schemas/sample.json [hunk 2]
```diff
@@ -12,16 +12,13 @@
             "type": "string",
             "description": "Sample name"
         },
-        "projects": {
```
Comment (danielballan) [inline amostra/schemas/sample.json]: I think we want to keep this field, but instead of being a list of foreign keys to documents/objects in amostra, it will be a lit of references to _external_ systems such as PASS. Thus, for the purposes of this PR, projects is any string. (We might eventually want to restrict that it should be a string formatted like as URL, as in `scheme://path`.

File: amostra/schemas/sample.json [hunk 3]
```diff
@@ -12,16 +12,13 @@
...
-            "type": "array",
-            "description": "List of foreign keys to Projects"
-        },
         "composition": {
             "type": "string",
             "description": "Sample composition (IUCr specification)"
         },
         "tags": {
             "type": "array",
+            "items": {"type": "string"},
```
Comment (danielballan) [inline amostra/schemas/sample.json]: :+1:

File: amostra/tests/test_jsonschema.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+from hypothesis import strategies as st
+
+from amostra.utils import load_schema
+
+sample_dict = load_schema("sample.json")
+# Pop uuid and revision cause they are created automatically
+sample_dict['properties'].pop('uuid')
+sample_dict['properties'].pop('revision')
+sample_dict['required'] = ['name']
+sample_dict['properties']['name']['minLength'] = 3
```
Comment (danielballan) [inline amostra/tests/test_jsonschema.py]: Are we _required_ by hypothesis to specific a range of allowed lengths? What does it do if we leave this unset?

On the other hand, it seems reasonable to add a `minLength` to the `sample.json` schema itself --- that is, maybe we shouldn't allow the user to name something `''`. And maybe a very large `maxLength` would be reasonable too.

Whatever we choose to do, I think it would be best if the schema we generate from is as close to the actual schema as possible.

File: amostra/tests/test_jsonschema.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+sample_dict['properties']['name']['maxLength'] = 5
+sample_dict['additionalProperties'] = False
+st_sample = hypothesis_jsonschema.from_schema(sample_dict)
```
Comment (danielballan) [inline amostra/tests/test_jsonschema.py]: I think this should be set in the schema, same as you did for Container.

File: amostra/tests/test_jsonschema.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+
+container_dict = load_schema("container.json")
+container_dict['properties'].pop('uuid')
+container_dict['properties'].pop('revision')
+container_dict['required'] = ['name', 'kind', 'contents']
```
Comment (danielballan) [inline amostra/tests/test_jsonschema.py]: Same comments as above... it would be good if the _only_ change we had to make here was popping the read-only keys, `uuid` and `revision`.

File: amostra/tests/test_jsonschema.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+st_container = hypothesis_jsonschema.from_schema(container_dict)
+
+
+@given(samples_list=st.lists(st_sample, unique_by=lambda x: x['name'], min_size=3, max_size=5),
+       containers_list=st.lists(st_container, unique_by=lambda x: x['name'], min_size=3, max_size=5))
+@settings(max_examples=3)
+def test_new(client, samples_list, containers_list):
+    contents_dict = dict()
+    for sample in samples_list:
+        assert client.samples.new(**sample)
```
Comment (danielballan) [inline amostra/tests/test_jsonschema.py]: What is this `assert` meant to verify?

File: amostra/schemas/container.json (hunk 1 above)
Comment (ke-zhang-rd) [reply amostra/schemas/container.json]: In code here, both kind and content needed.
https://github.com/NSLS-II/amostra/blob/master/amostra/objects.py#L175

In design, 
contents needed make sense, not sure about kind

Comment (ke-zhang-rd) [reply amostra/schemas/container.json]: it doesn't have really strong connection with this PR. I updated them to make fake container works(has kind and contents).

File: amostra/tests/test_jsonschema.py (hunk 7 above)
Comment (ke-zhang-rd) [reply amostra/tests/test_jsonschema.py]: Successfully created new sample. 
Maybe there should be pipeline test to actually use these fake sample and container and assert later.

fake sample/container -> create new object -> update -> update -> revert -> ... -> othermethod call -> assert expect == actual

File: amostra/tests/test_jsonschema.py (hunk 4 above)
Comment (ke-zhang-rd) [reply amostra/tests/test_jsonschema.py]: specify a range is optional. 

I did this to get reasonable length. Cause without them, hypothesis always try to catch edge case at very beginning. What I mean is edge cases like '' and extremely long string have higher possibility to appear in a random list of strings.

Actually, I put them in test instead of in json file is notice from group meeting that how ambitious scientists is in naming and I don't want to stop them by a range. I try to limit this range locally(in test only).

I don't have strong opinion about this. Does  minLength = 1, maxLength = 50 sound good in json file?

File: amostra/tests/test_jsonschema.py (hunk 5 above)
Comment (ke-zhang-rd) [reply amostra/tests/test_jsonschema.py]: Ok

File: amostra/tests/test_jsonschema.py (hunk 6 above)
Comment (ke-zhang-rd) [reply amostra/tests/test_jsonschema.py]: Same reason here. https://github.com/NSLS-II/amostra/pull/44/files#r319613117

File: amostra/tests/test_jsonschema.py (hunk 4 above)
Comment (danielballan) [reply amostra/tests/test_jsonschema.py]: If it's optional, let's leave it unconstrained for now. Why should we mind if hypothesis tests edge cases? That seems good.

File: amostra/schemas/container.json (hunk 1 above)
Comment (danielballan) [reply amostra/schemas/container.json]: OK, then let's leave the schema as is. Schema changes should be motivated by user needs, not testing.

File: amostra/tests/test_jsonschema.py (hunk 7 above)
Comment (danielballan) [reply amostra/tests/test_jsonschema.py]: I think this might be misunderstanding what `assert` does. The line below this one,

```python
s = client.samples.new(**sample)
```
is sufficient to check that sample creation happened without raising any errors. If it does raise any errors, the test will fail. No `assert` necessary.

The statement `assert client.samples.new(**sample)` that `client.samples.new(**sample)` returns something that is _true_, which is superfluous.

File: amostra/tests/test_jsonschema.py (hunk 6 above)
Comment (danielballan) [reply amostra/tests/test_jsonschema.py]: I think changing the schema before testing cuts against the spirit of how to use hypothesis. We _have_ to dispense with `uuid` and `required` become of how our API works, but it would be better not to mutate anything else about the schema.

File: amostra/tests/test_jsonschema.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+from hypothesis import strategies as st
+
+from amostra.utils import load_schema
+
+sample_dict = load_schema("sample.json")
+# Pop uuid and revision cause they are created automatically
+sample_dict['properties'].pop('uuid')
+sample_dict['properties'].pop('revision')
+# sample_dict['properties']['name']['minLength'] = 1
+sample_dict['required'] = sample_dict['required'][2:]
```
Comment (danielballan) [inline amostra/tests/test_jsonschema.py]: I think it would be more readable and robust to explicitly `remove` them, rather than implicitly relying and `uuid` and `revision` being the first two items in the list.

File: amostra/tests/test_jsonschema.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+from hypothesis import HealthCheck, given, settings
+from hypothesis import strategies as st
+
+from amostra.utils import load_schema
+
+sample_dict = load_schema("sample.json")
+# Pop uuid and revision cause they are created automatically
+sample_dict['properties'].pop('uuid')
+sample_dict['properties'].pop('revision')
+# sample_dict['properties']['name']['minLength'] = 1
```
Comment (danielballan) [inline amostra/tests/test_jsonschema.py]: This line can just be deleted.

File: amostra/tests/test_jsonschema.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+sample_dict['required'] = sample_dict['required'][2:]
+st_sample = hypothesis_jsonschema.from_schema(sample_dict)
+
+container_dict = load_schema("container.json")
+container_dict['properties'].pop('uuid')
+container_dict['properties'].pop('revision')
+# container_dict['properties']['name']['minLength'] = 1
```
Comment (danielballan) [inline amostra/tests/test_jsonschema.py]: Same as above.

**Comment by ke-zhang-rd** [comment]
```
During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
~/miniconda3/envs/py3/lib/python3.7/site-packages/traitlets/traitlets.py in get(self, obj, cls)
    527         try:
--> 528             value = obj._trait_values[self.name]
    529         except KeyError:

KeyError: 'projects'

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
~/miniconda3/envs/py3/lib/python3.7/site-packages/traitlets/traitlets.py in get(self, obj, cls)
    527         try:
--> 528             value = obj._trait_values[self.name]
    529         except KeyError:

KeyError: 'projects'

During handling of the above exception, another exception occurred:

RecursionError                            Traceback (most recent call last)
~/amostra/amostra/revert_test.py in <module>
      5 db_name = str(uuid.uuid4())
      6 client = amostra.mongo_client.Client('mongodb://localhost:27017/' + db_name)
----> 7 s = client.samples.new(name = '')

~/amostra/amostra/mongo_client.py in new(self, *args, **kwargs)
    147 
    148     def new(self, *args, **kwargs):
--> 149         return self._client._new_document(self._obj_type, args, kwargs)
    150 
    151     def find(self, filter):

~/amostra/amostra/mongo_client.py in _new_document(self, obj_type, args, kwargs)
     59 
     60         # Insert the new object.
---> 61         collection.insert_one(obj.to_dict())
     62 
     63         # Observe any updates to the object and sync them to MongoDB.

~/amostra/amostra/objects.py in to_dict(self)
     57         Represent the object as a JSON-serializable dictionary.
     58         """
---> 59         return {name: getattr(self, name) for name in self.trait_names()}
     60 
     61     @classmethod

~/amostra/amostra/objects.py in <dictcomp>(.0)
     57         Represent the object as a JSON-serializable dictionary.
     58         """
---> 59         return {name: getattr(self, name) for name in self.trait_names()}
     60 
     61     @classmethod

~/miniconda3/envs/py3/lib/python3.7/site-packages/traitlets/traitlets.py in __get__(self, obj, cls)
    554             return self
    555         else:
--> 556             return self.get(obj, cls)
    557 
    558     def set(self, obj, value):

~/miniconda3/envs/py3/lib/python3.7/site-packages/traitlets/traitlets.py in get(self, obj, cls)
    533                 raise TraitError("No default value found for %s trait of %r"
    534                                  % (self.name, obj))
--> 535             value = self._validate(obj, dynamic_default())
    536             obj._trait_values[self.name] = value
    537             return value

~/miniconda3/envs/py3/lib/python3.7/site-packages/traitlets/traitlets.py in _validate(self, obj, value)
    591             value = self.validate(obj, value)
    592         if obj._cross_validation_lock is False:
--> 593             value = self._cross_validate(obj, value)
    594         return value
    595 

~/miniconda3/envs/py3/lib/python3.7/site-packages/traitlets/traitlets.py in _cross_validate(self, obj, value)
    597         if self.name in obj._trait_validators:
    598             proposal = Bunch({'trait': self, 'value': value, 'owner': obj})
--> 599             value = obj._trait_validators[self.name](obj, proposal)
    600         elif hasattr(obj, '_%s_validate' % self.name):
    601             meth_name = '_%s_validate' % self.name

~/miniconda3/envs/py3/lib/python3.7/site-packages/traitlets/traitlets.py in __call__(self, *args, **kwargs)
    905         """Pass `*args` and `**kwargs` to the handler's function if it exists."""
    906         if hasattr(self, 'func'):
--> 907             return self.func(*args, **kwargs)
    908         else:
    909             return self._init_call(*args, **kwargs)

~/amostra/amostra/objects.py in _validate_with_jsonschema(instance, proposal)
     22     This is meant to be used with traitlets' @validate decorator.
     23     """
---> 24     jsonschema.validate(instance.to_dict(), instance.SCHEMA)
     25     return proposal['value']
     26 

... last 8 frames repeated, from the frame below ...

~/amostra/amostra/objects.py in to_dict(self)
     57         Represent the object as a JSON-serializable dictionary.
     58         """
---> 59         return {name: getattr(self, name) for name in self.trait_names()}
     60 
     61     @classmethod

RecursionError: maximum recursion depth exceeded

```

The error was observed when try to init a `Sample` with only empty string `name`

```
from pymongo import MongoClient
import uuid
import amostra.mongo_client

db_name = str(uuid.uuid4())
client = amostra.mongo_client.Client('mongodb://localhost:27017/' + db_name)
s = client.samples.new(name='') 
```

The reason isn't clear yet. I suspect it come from `traitlets`'s `Unicode`. 
Right now, in sample.json, set `"minLength": 1` to go around this error.

**Comment by danielballan** [comment]
Interesting. I think it would be good to understand the cause before merging. Can you start with something minimal like:

```py
class Thing(amostra.objects.AmostraDocument): 
    stuff = traitlets.List(traitlets.Unicode()) 
```

**Comment by ke-zhang-rd** [comment]
> ```python
> class Thing(amostra.objects.AmostraDocument): 
> ```

I'll try. Do you have clue/direction why `name`(which is `Unicode()`) empty or not could influence `project` which is `List(Unicode())` behavior?

**Comment by ke-zhang-rd** [comment]
After some search online, I feel typical way to use __new__ is
```
def __new__(cls, *args, **kwargs):
    inst = super().__new__(cls)
    '''
    manipulate inst
    '''
    return inst
```

Also in traitlets examples [here](https://traitlets.readthedocs.io/en/stable/using_traitlets.html), looks `validate` method bond with instance instead of class.

```
class Parity(HasTraits):
    value = Int()
    parity = Int()

    @validate('value')
    def _valid_value(self, proposal):
```

File: amostra/objects.py [hunk 11]
```diff
@@ -39,8 +39,9 @@ def __init__(self, _amostra_client, *args, **kwargs):
     def __new__(cls, *args, **kwargs):
         # Configure _validate_with_jsonschema to validate all traits.
         trait_names = list(cls.class_traits())
-        cls._validate = validate(*trait_names)(_validate_with_jsonschema)
-        return super().__new__(cls, *args, **kwargs)
+        instance = super().__new__(cls, *args, **kwargs)
+        instance._validate = validate(*trait_names)(_validate_with_jsonschema)
```
Comment (tacaswell) [inline amostra/objects.py]: I think think that this change makes the validation not happen at al!  The `validate` method produces a descriptor which needs to be in place for the `super().__new__(...)` to find them and do something about it.

Are there any tests we we are sure that we do reject invalid documents?

Comment (danielballan) [reply amostra/objects.py]: Yes, I think @tacaswell is right. It would be good to add a test that uses `pytest.raises` to ensure that validation is running and correctly raising an error on invalid inputs.

Comment (ke-zhang-rd) [reply amostra/objects.py]: I agree with you that validation wasn't triggered.

**Comment by tacaswell** [comment]
```
Inlt [24]: s                                                                                                                                                                                                                                                    
Out[24]: amostra.objects.Sample

In [25]: s(None, name='') 
```

is enough to trigger the recursion.  I suspect this is due to `''` being the default value of `name` and it looks like there is a loop triggered via the interplay interplay between the collective validation during setting the values and during getting the default values. 

Given that we have already made 'name' special via the signature, I think we should enforce that it is not the empty sttring in the `Sample` init.

**Comment by ke-zhang-rd** [comment]
I suspect [this](https://github.com/NSLS-II/amostra/blob/179d0f7a8192ccfada2559ef8a0cb9226f72b6cc/amostra/http_client.py#L60) need to be fixed?

short words about why recursion
`to_dict` -> `_validate_with_jsonschema` -> `to_dict` ...

**Comment by ke-zhang-rd** [comment]
Maybe I should do thing below instead of every place using `to_dict`.
```
    def to_dict(self):
        """
        Represent the object as a JSON-serializable dictionary.
        """
        with self. cross_validation_lock:
            result = {name: getattr(self, name) for name in self.trait_names()}
        return result
```

**Comment by danielballan** [comment]
Great, using the lock inside `to_dict` (and removing it from `__repr__` and elsewhere) feels right. I will review again with fresh eyes next week. This is too detailed for a Friday 5pm review. :-D

File: amostra/schemas/container.json (hunk 1 above)
Comment (danielballan) [reply amostra/schemas/container.json]: I don't understand why you marked this as "resolved".

File: amostra/tests/test_jsonschema.py [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+container_dict = load_schema("container.json")
+container_dict['properties'].pop('uuid')
+container_dict['properties'].pop('revision')
+container_dict['required'].remove('uuid')
+container_dict['required'].remove('revision')
+st_container = hypothesis_jsonschema.from_schema(container_dict)
+
+
+@given(samples_list=st.lists(st_sample, unique_by=lambda x: x['name'], min_size=3, max_size=5),
```
Comment (danielballan) [inline amostra/tests/test_jsonschema.py]: You might consider [`operator.itemgetter`](https://docs.python.org/3/library/operator.html#operator.itemgetter) here, but a `lambda` is also OK.
